src/streams.c

```diff
@@ -79,69 +81,125 @@ static Int READ_COMMAND(Obj *evalResult)
+/****************************************************************************
+**
+*F  FuncREAD_ALL_COMMANDS( <self>, <instream>, <echo>, <outputFunc> )
+**
+**  FuncREAD_ALL_COMMANDS reads all code from <instream> and returns either
+**  fail if the stream cannot be opened, or executes all statements in the
+**  stream and returns a list of lists, each entry of which reflects the
+**  result of the execution of one statement.
+**
+**  The results are returned as lists of length three if <outputFunc> is
+**  false, or five, if <outputFunc> is a function.
+**
+**  If the first entry is true, then the second entry is bound to the result
+**  of the statement if there was one, and unbound otherwise, and the third
+**  entry reflects whether the statement ended in a dual semicolon. If the
+**  first entry is false, an error occurred during the execution of a
+**  statement. In this case, if <outputFunc> is false, only the first entry
+**  is set. If <outputFunc> is a function, only the first and the fifth entry
+**  are set. If <outputFunc> is a function, the fourth entry of the list is
+**  <outputFunc> applied to the result, if there was one, and if the
+**  statement did not end in a dual semicolon. Furthermore, if <outputFunc>
+**  is not false, the fifth entry contains a string of all output printed
+**  while the corresponding statement was executed. If <outputFunc> is false,
+**  all intermediate output is discarded.
+**
+**  This function is currently used in interactive tools such as the GAP
+**  Jupyter kernel to execute cells and is likely to be replaced by a
+**  function that can read a single command from a stream without losing the
+**  rest of its content.
+*/
+Obj FuncREAD_ALL_COMMANDS(Obj self, Obj instream, Obj echo, Obj outputFunc)
 {
     ExecStatus status;
-    Int resultCount, resultCapacity;
-    UInt dualSemicolon;
-    Obj result, resultList;
-    Obj evalResult;
+    Int        resultCount;
+    UInt       dualSemicolon;
+    Obj        result, resultList;
+    Obj        evalResult;
+    Obj        outstream;
+    Obj        outstreamString;
+    Obj        outstreamStringCopy;
+    Obj        outputFunc_return;
+
+    /* try to open the streams */
+    if (!OpenInputStream(instream, echo == True)) {
+        return Fail;
+    }
 
-    /* try to open the stream */
-    if (!OpenInputStream(stream, echo == True)) {
+
+    if (outputFunc != False) {
+        outstreamString = NEW_STRING(0);
+        outstream = DoOperation2Args(ValGVar(GVarName("OutputTextString")),
+                                     outstreamString, True);
+    }
+    else {
+        outstreamString = False;
+        outstream = False;
+    }
+    if (outstream != False && !OpenOutputStream(outstream)) {
+        CloseInput();
         return Fail;
     }
 
     resultCount = 0;
-    resultCapacity = 16;
-    resultList = NEW_PLIST( T_PLIST, resultCapacity );
+    resultList = NEW_PLIST(T_PLIST, 16);
```

Comment:
I took the liberty of fixing the formatting of your comment for `FuncREAD_ALL_COMMANDS`.

But now I see that you dropped most if not all of the other changes I had added previously to your PR (via pull request that you merged). Is there a reason for that?